--- ./unzip.c.orig	2018-06-01 15:01:57.666547940 +0800
+++ ./unzip.c.new	2018-06-01 15:02:10.518547340 +0800
@@ -108,6 +108,16 @@
 #define UNZ_MAXFILENAMEINZIP (256)
 #endif
 
+static unz_memfunc g_unz_memapi = {0, 0};
+
+extern int ZEXPORT unz_set_memfuncs(const unz_memfunc* memfunc)
+{
+	g_unz_memapi = *memfunc;
+}
+
+#define ALLOC(size)  ((*(g_unz_memapi.m_malloc))(size))
+#define TRYFREE(p)   { if (p) { (*(g_unz_memapi.m_free))(p); p = NULL; } }
+
 #ifndef ALLOC
 # define ALLOC(size) (malloc(size))
 #endif
