diff -uprN libcgroup-0.41/src/daemon/cgrulesengd.c libcgroup-0.41-patch/src/daemon/cgrulesengd.c
--- libcgroup-0.41/src/daemon/cgrulesengd.c	2019-05-28 09:55:37.000000000 +0800
+++ libcgroup-0.41-patch/src/daemon/cgrulesengd.c	2019-12-12 17:47:07.936407318 +0800
@@ -889,9 +889,6 @@ int cgre_start_daemon(const char *logp,
 		} else if (pid > 0) {
 			exit(EXIT_SUCCESS);
 		}
-
-		/* Change the file mode mask. */
-		umask(0);
 	} else {
 		flog(LOG_DEBUG, "Not using daemon mode\n");
 		pid = getpid();
